File: engine/formation.py
import math


class Formation:
    """
    Genera posiciones para movimiento en formación.
    """

    @staticmethod
    def square(target_x, target_y, count, spacing=1):
        positions = []

        if count <= 0:
            return positions

        size = math.ceil(math.sqrt(count))

        start_x = target_x - size // 2
        start_y = target_y - size // 2

        for i in range(count):
            x = start_x + (i % size) * spacing
            y = start_y + (i // size) * spacing
            positions.append((int(x), int(y)))

        return positions

    @staticmethod
    def line(target_x, target_y, count, spacing=1):
        positions = []

        start_x = target_x - count // 2

        for i in range(count):
            positions.append((int(start_x + i * spacing), int(target_y)))

        return positions

    @staticmethod
    def column(target_x, target_y, count, spacing=1):
        positions = []

        start_y = target_y - count // 2

        for i in range(count):
            positions.append((int(target_x), int(start_y + i * spacing)))

        return positions

    @staticmethod
    def circle(target_x, target_y, count, radius=2):
        positions = []

        if count <= 0:
            return positions

        for i in range(count):
            angle = (math.pi * 2 * i) / count
            x = target_x + math.cos(angle) * radius
            y = target_y + math.sin(angle) * radius
            positions.append((int(round(x)), int(round(y))))

        return positions

    @staticmethod
    def create(kind, target_x, target_y, count, spacing=1):
        if kind == "line":
            return Formation.line(target_x, target_y, count, spacing)

        if kind == "column":
            return Formation.column(target_x, target_y, count, spacing)

        if kind == "circle":
            return Formation.circle(target_x, target_y, count, max(2, spacing * 2))

        return Formation.square(target_x, target_y, count, spacing)

File: src/engine/formation.rs
use std::f64::consts::TAU;

pub struct Formation;

impl Formation {
    pub fn positions(
        kind: &str,
        count: usize,
        center: (f64, f64),
        spacing: f64,
    ) -> Vec<(f64, f64)> {
        match kind {
            "line" => (0..count)
                .map(|i| {
                    (
                        center.0 + Self::centered_offset(i, count) * spacing,
                        center.1,
                    )
                })
                .collect(),
            "column" => (0..count)
                .map(|i| {
                    (
                        center.0,
                        center.1 + Self::centered_offset(i, count) * spacing,
                    )
                })
                .collect(),
            "staggered" => (0..count)
                .map(|i| {
                    let row = i / 2;
                    let side = if i % 2 == 0 { -0.5 } else { 0.5 };
                    (
                        center.0 + side * spacing,
                        center.1 + Self::centered_offset(row, count.div_ceil(2)) * spacing,
                    )
                })
                .collect(),
            "wedge" => (0..count)
                .map(|i| {
                    if i == 0 {
                        return center;
                    }
                    let row = ((i + 1) as f64).sqrt().floor() as usize;
                    let side = if i % 2 == 0 { -1.0 } else { 1.0 };
                    (
                        center.0 + side * row as f64 * spacing * 0.5,
                        center.1 + row as f64 * spacing,
                    )
                })
                .collect(),
            "circle" => (0..count)
                .map(|i| {
                    let angle = i as f64 / count.max(1) as f64 * TAU;
                    (
                        center.0 + angle.cos() * spacing,
                        center.1 + angle.sin() * spacing,
                    )
                })
                .collect(),
            _ => {
                let side = (count as f64).sqrt().ceil() as usize;
                let rows = count.div_ceil(side.max(1));
                (0..count)
                    .map(|i| {
                        let x = i % side;
                        let y = i / side;
                        (
                            center.0 + Self::centered_offset(x, side) * spacing,
                            center.1 + Self::centered_offset(y, rows) * spacing,
                        )
                    })
                    .collect()
            }
        }
    }

    fn centered_offset(index: usize, count: usize) -> f64 {
        if count <= 1 {
            0.0
        } else {
            index as f64 - (count.saturating_sub(1) as f64 * 0.5)
        }
    }
}
